sql/src/main/java/io/crate/protocols/postgres/BulkPortal.java

```diff
@@ -0,0 +1,177 @@
+import io.crate.action.sql.ResultReceiver;
+import io.crate.analyze.Analysis;
+import io.crate.analyze.Analyzer;
+import io.crate.analyze.ParameterContext;
+import io.crate.analyze.symbol.Field;
+import io.crate.analyze.symbol.Symbols;
+import io.crate.concurrent.CompletionListener;
+import io.crate.exceptions.Exceptions;
+import io.crate.exceptions.UnsupportedFeatureException;
+import io.crate.executor.Executor;
+import io.crate.executor.TaskResult;
+import io.crate.operation.collect.StatsTables;
+import io.crate.planner.Plan;
+import io.crate.planner.Planner;
+import io.crate.sql.tree.Statement;
+import io.crate.types.DataType;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+
+import static io.crate.action.sql.SQLRequest.EMPTY_ARGS;
+
+class BulkPortal extends AbstractPortal {
+
+    private final List<List<Object>> bulkParams = new ArrayList<>();
+    private final List<ResultReceiver> resultReceivers = new ArrayList<>();
+    private String query;
+    private Statement statement;
+    private Analysis analysis;
+    private Plan plan;
+    private int maxRows = 0;
+    private List<? extends DataType> outputTypes;
+
+    BulkPortal(String name,
+               String query,
+               Statement statement,
+               Analysis analysis,
+               Plan plan,
+               List<? extends DataType> outputTypes,
+               ResultReceiver resultReceiver,
+               int maxRows,
+               List<Object> params,
+               SessionData sessionData) {
+        super(name, sessionData);
+        this.query = query;
+        this.statement = statement;
+        this.analysis = analysis;
+        this.plan = plan;
+        this.outputTypes = outputTypes;
+        this.resultReceivers.add(resultReceiver);
+        this.maxRows = maxRows;
+        this.bulkParams.add(params);
+    }
+
+    @Override
+    public FormatCodes.FormatCode[] getResultFormatCodes() {
+        return new FormatCodes.FormatCode[0];
+    }
+
+    @Override
+    public List<? extends DataType> getOutputTypes() {
+        return outputTypes;
+    }
+
+    @Override
+    public String getLastQuery() {
+        return this.query;
+    }
+
+    @Override
+    public Portal bind(String statementName,
+                       String query,
+                       Statement statement,
+                       List<Object> params,
+                       @Nullable FormatCodes.FormatCode[] resultFormatCodes) {
+        this.bulkParams.add(params);
+        return this;
+    }
+
+    @Override
+    public List<Field> describe() {
+        return null;
+    }
+
+    @Override
+    public void execute(ResultReceiver resultReceiver, int maxRows) {
+        if (maxRows != 1) {
+            throw new UnsupportedFeatureException("bulk operations don't support fetch size");
+        }
+        validateReadOnly(analysis);
```

Comment:
I think instead of using `validateReadOnly` could be removed from the abstract class and instead be more specific to the portal.
In BulkPortal for example only works for write operations